aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu

```diff
@@ -708,4 +714,271 @@ Tensor _sparse_sum_backward_cuda(const Tensor& grad_, const SparseTensor& input_
+  ) {
+    bool trim_right = mid_val > target_mat_num;
+    int64_t mid_idx_minus_1 = mid_idx - 1;
+    int64_t mid_idx_plus_1 = mid_idx + 1;
+
+    end_idx = trim_right ? mid_idx_minus_1 : end_idx;
+    start_idx = trim_right ? start_idx : mid_idx_plus_1;
+    mid_idx = (start_idx + end_idx) >> 1;
+    mid_val = indices_1D[mid_idx*indices_1D_stride];
+  }
+
+  found = (mid_val == target_mat_num)
+    && (
+      (mid_idx == (num_elements-1))
+      || (indices_1D[(mid_idx+1)*indices_1D_stride] != target_mat_num)
+    );
+
+  mat_el_end_indices[target_mat_num] = found ? mid_idx : -1;
+}
+
+// Search through a 1D tensor of sorted sparse matrix
+// indices to find the end index for each matrix
+void search_end_matrix_indices(int64_t* mat_el_end_indices, int64_t num_matrices, const LongTensor& indices_1D) {
+  int curDevice = -1;
+  cudaGetDevice(&curDevice);
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream(curDevice);
+
+  auto indices_1D_ti = getTensorInfo<int64_t, int64_t>(indices_1D);
+  int64_t grid_size = (num_matrices / 64)+1;
+  int64_t block_size = 64;
+  int64_t num_elements = indices_1D.size(0);
+
+  search_end_matrix_indices_cuda_kernel<<<grid_size, block_size, 0, stream>>>(
+    mat_el_end_indices,
+    num_matrices,
+    indices_1D_ti,
+    num_elements
+  );
+  cudaDeviceSynchronize();
+}
+
+cudaDataType getTensorCudaDataType(Tensor self) {
+  cudaDataType cuda_data_type;
+  switch (self.scalar_type()) {
+    case ScalarType::Float:
+      cuda_data_type = CUDA_R_32F;
+      break;
+    case ScalarType::Double:
+      cuda_data_type = CUDA_R_64F;
+      break;
+    default:
+      TORCH_CHECK(false, "Tensor types must be either float32 or float64");
+      break;
+  }
+  return cuda_data_type;
+}
+
+Tensor& bmm_out_sparse_cuda(Tensor& result, const SparseTensor& self, const Tensor& mat2) {
+  return bmm_out_sparse_cuda(result, self, mat2, false);
+}
+
+Tensor& bmm_out_sparse_cuda(Tensor& result, const SparseTensor& self, const Tensor& mat2, bool deterministic) {
+  TORCH_CHECK(!mat2.is_sparse(), "bmm_sparse: Tensor 'mat2' must be dense");
+  TORCH_CHECK(self.dense_dim() == 0, "bmm_sparse: Tensor 'self' must have 0 dense dims, but has ", self.dense_dim());
+  TORCH_CHECK(self.sparse_dim() == 3, "bmm_sparse: Tensor 'self' must have 3 sparse dims, but has ", self.sparse_dim());
+  TORCH_CHECK(mat2.dim() == 3, "bmm_sparse: Tensor 'mat2' must have 3 dims, but has ", mat2.dim());
+  TORCH_CHECK(self.size(0) == mat2.size(0), "bmm_sparse: 'self.size(0)' and 'mat2.size(0)' must match");
+  TORCH_CHECK(self.size(2) == mat2.size(1), "bmm_sparse: 'self.size(2)' and 'mat2.size(1)' must match");
+
+  result.resize_({self.size(0), mat2.size(2), self.size(1)});
+
+  if ((self._nnz() == 0) || (mat2.size(1) == 0) || (mat2.size(2) == 0)) {
+    result.zero_().transpose_(1,2);
+    return result;
+  }
+
+  // TODO: I'm not sure if this is necessary or not. Does at::resize_()
+  //    ensure contiguity?
+  if (!result.is_contiguous()) {
```

Comment:
Does anyone know if `at::resize_()` ensures contiguity or not? If it does, I can remove this check.

Reply:
No, `resize_` does not ensure contiguity. Specifically, if you have a non-contiguous tensor of exactly the right size, `resize_` will no op. However, please also note that if you are implementing the `out=` variant of a function, you cannot just call contiguous on it, that will not work. You will need to do the computation into a contiguous out buffer, then do a copy into the non-contiguous preallocated memory.

Reply:
Ah ok, I'll do that.

Reply:
Alright, what I chose to do is use a separate temporary buffer only if the given one is not contiguous. If it is already contiguous, then we can avoid the performance hit of having to copy the result.

Reply:
Yep, that sounds right to me.